sentry-android-core/src/main/java/io/sentry/android/core/AndroidTraceTransactionListener.java

```diff
@@ -0,0 +1,150 @@
+      return;
+    }
+    traceFilesDir = new File(tracesFilesDirPath);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public synchronized void onTransactionStart(ITransaction transaction) {
+
+    // Debug.startMethodTracingSampling() is only available since Lollipop
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) return;
+
+    // Let's be sure to end any running trace
+    if (activeTransaction != null) onTransactionEnd(activeTransaction);
+
+    traceFile = FileUtils.resolve(traceFilesDir, "sentry-" + UUID.randomUUID() + ".trace");
+
+    if (traceFile == null) {
+      options.getLogger().log(SentryLevel.DEBUG, "Could not create a trace file");
+      return;
+    }
+
+    if (traceFile.exists()) {
+      options
+          .getLogger()
+          .log(SentryLevel.DEBUG, "Trace file already exists: %s", traceFile.getPath());
+      return;
+    }
+
+    long intervalMs = options.getProfilingTracesIntervalMillis();
+    if (intervalMs <= 0) {
+      options
+          .getLogger()
+          .log(SentryLevel.DEBUG, "Profiling trace interval is set to %d milliseconds", intervalMs);
+      return;
+    }
+
+    // We stop the trace after 30 seconds, since such a long trace is very probably a trace
+    // that will never end due to an error
+    SentryExecutors.tracingExecutor.schedule(
+        () -> onTransactionEnd(transaction), 30_000, MILLISECONDS);
+
+    startedMethodTracing = true;
+    int intervalUs = (int) MILLISECONDS.toMicros(intervalMs);
+    activeTransaction = transaction;
+    Debug.startMethodTracingSampling(traceFile.getPath(), BUFFER_SIZE_BYTES, intervalUs);
+  }
+
+  @Override
+  public synchronized void onTransactionEnd(ITransaction transaction) {
+    // In case a previous timeout tries to end a newer transaction we simply ignore it
+    if (transaction != activeTransaction) return;
+
+    if (startedMethodTracing) {
+      startedMethodTracing = false;
+      activeTransaction = null;
+
+      Debug.stopMethodTracing();
+
+      if (traceFile == null || !traceFile.exists()) {
+        options
+            .getLogger()
+            .log(
+                SentryLevel.DEBUG,
+                "Trace file %s does not exists",
+                traceFile == null ? "null" : traceFile.getPath());
+        return;
+      }
+
+      // todo should I use transaction.getEventId() instead of new SentryId()?
+      //  Or should I add the transaction id as an header to the envelope?
+      //  Or should I simply ignore the transaction entirely (wouldn't make any sense)?
+      //  And how to check if a trace is from a startup?
+      try {
+        SentryEnvelope envelope =
+            SentryEnvelope.from(
+                new SentryId(),
+                traceFile.getPath(),
+                traceFile.getName(),
+                options.getSdkVersion(),
+                options.getMaxTraceFileSize(),
+                true);
+        hub.captureEnvelope(envelope);
```

Comment:
The transaction already calls `captureEnvelope` with the transaction item, ideally, this traceFile would be one more item of that envelope.

Reply:
I'm adding it to the same envelope. Just one thing: 
Can I add 2 fields in the `SentryEnvelopeItemHeader`? We are thinking to add a transaction id and the span id the profiling trace refers to.
Of course, I would add the needed logic in `SentryEnvelopeItemHeaderAdapter`, too

Reply:
This should be discussed with the ingest team, if they say that these fields go into the Item header of the envelope, then sure.

Reply:
After several discussions we will go with a json payload with all needed variables in it, without any custom headers